classes/task/collect_metrics_task.php

```diff
@@ -0,0 +1,136 @@
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+namespace tool_cloudmetrics\task;
+
+use tool_cloudmetrics\metric;
+use tool_cloudmetrics\collector;
+
+/**
+ * Controls the running of metrics within a single task. Derive from scheduled_task to enable running from
+ * the moodle task system, but it is  intendend to be run separately.
+ *
+ * @package   tool_cloudmetrics
+ * @author    Jason den Dulk <jasondendulk@catalyst-au.net>
+ * @copyright 2022, Catalyst IT
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+class collect_metrics_task extends \core\task\scheduled_task {
+
+    const FIVEMINUTES = 5;
+    const FIFTEENMINUTES = 15;
+    const THIRTYMINUTES = 30;
+    const ONEHOUR = 60;
+    const THREEHOURS = 180;
+    const TWELVEHOURS = 720;
+    const ONEDAY = 1440;
+    const ONEWEEK = 10080;
+
+    private $time = null;
+
+    /**
+     * Display name for the task name.
+     *
+     * @return \lang_string|string
+     * @throws \coding_exception
+     */
+    public function get_name() {
+        return get_string('task_runner', 'tool_cloudmetrics');
+    }
+
+    /**
+     * Set a specific time for the use. Defaults to now.
+     *
+     * @param int $time
+     */
+    public function set_time(int $time) {
+        $this->time = $time;
+    }
+
+    /**
+     * Will send the metrics to the collectors.
+     *
+     * @param array $items
+     */
+    public function send_metrics(array $items) {
+        collector\manager::send_metrics($items);
+    }
+
+    /**
+     * Determine which level of frequency is desired for the metrics.
+     *
+     * @param $timediff
+     * @return int
+     */
+    public function get_frequency_cutoff($timediff): int {
+        if ($timediff % self::ONEWEEK == 0) {
+            return metric\manager::FREQ_WEEK;
+        } else if ($timediff % self::ONEDAY == 0) {
+            return metric\manager::FREQ_DAY;
+        } else if ($timediff % self::TWELVEHOURS == 0) {
+            return metric\manager::FREQ_12HOUR;
+        } else if ($timediff % self::THREEHOURS == 0) {
+            return metric\manager::FREQ_3HOUR;
+        } else if ($timediff % self::ONEHOUR == 0) {
+            return metric\manager::FREQ_HOUR;
+        } else if ($timediff % self::THIRTYMINUTES == 0) {
+            return metric\manager::FREQ_30MIN;
+        } else if ($timediff % self::FIFTEENMINUTES == 0) {
+            return metric\manager::FREQ_15MIN;
+        } else if ($timediff % self::FIVEMINUTES == 0) {
+            return metric\manager::FREQ_5MIN;
+        } else {
+            return metric\manager::FREQ_MIN;
+        }
+    }
+
+    /**
+     * Perform the task
+     *
+     * @throws \Exception
+     */
+    public function execute() {
+        // This algorithm is performance important. Any opportunity to optimize should be welcome.
+
+        $tz = \core_date::get_server_timezone_object();
```

Comment:
This chunk of 10-15 lines I'd like a couple comments explaining the what and why of it, tz and time maths is never trivial

Reply:
I added a lot of comments to explain the algorithm used here.